File: code/client/core/source/operation-graph/generate-result.cpp
// <copyright file="generate-result.cpp" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

module;

#include <map>
#include <stdexcept>
#include <unordered_map>
#include <vector>

export module Soup.Core:GenerateResult;

import Opal;
import :CommandInfo;
import :OperationGraph;

using namespace Opal;

namespace Soup::Core 
{
	/// <summary>
	/// The generate result that represents the known operation graph that needs to 
	/// be evaluated and a boolean indication if this is a preprocessing run
	/// </summary>
	export class GenerateResult
	{
	private:
		OperationGraph _graph;
		bool _isPreprocessor;

	public:
		/// <summary>
		/// Initializes a new instance of the <see cref="GenerateResult"/> class.
		/// </summary>
		GenerateResult() :
			_graph(),
			_isPreprocessor()
		{
		}

		/// <summary>
		/// Initializes a new instance of the <see cref="GenerateResult"/> class.
		/// </summary>
		GenerateResult(
			OperationGraph graph,
			bool isPreprocessor) :
			_graph(std::move(graph)),
			_isPreprocessor(isPreprocessor)
		{
		}

		/// <summary>
		/// Get the operation graph
		/// </summary>
		const OperationGraph& GetGraph() const
		{
			return _graph;
		}

		/// <summary>
		/// Get the operation graph
		/// </summary>
		OperationGraph& GetGraph()
		{
			return _graph;
		}

		/// <summary>
		/// Get a value indicating if the graph is for preprocessing files
		/// </summary>
		bool IsPreprocessor() const
		{
			return _isPreprocessor;
		}
	};
}

File: code/generate-sharp/utilities/operation-graph/generate-result.cs
// <copyright file="generate-result.cs" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

namespace Soup.Build.Utilities;

/// <summary>
/// The cached operation graph that is used to track input/output mappings for previous build
/// executions to support incremental builds
/// </summary>
public class GenerateResult
{
	/// <summary>
	/// Initializes a new instance of the <see cref="GenerateResult"/> class.
	/// </summary>
	public GenerateResult()
	{
		this.EvaluateGraph = new OperationGraph();
		this.IsPreprocessor = false;
	}

	/// <summary>
	/// Initializes a new instance of the <see cref="GenerateResult"/> class.
	/// </summary>
	public GenerateResult(
		OperationGraph evaluateGraph,
		bool isPreprocessor)
	{
		this.EvaluateGraph = evaluateGraph;
		this.IsPreprocessor = isPreprocessor;
	}

	/// <summary>
	/// Get Evaluate Graph
	/// </summary>
	public OperationGraph EvaluateGraph { get; }

	/// <summary>
	/// Get a value indicating if the graph is preprocessing
	/// </summary>
	public bool IsPreprocessor { get; }
}